### ISSUE TITLE
vs_prediction_xgboost.py (전처리 데이터 바탕)

### ISSUE THREAD

**Issue by leeganghyun97** [KimSeungJun21/Team_project#2]
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import StackingRegressor
from sklearn.svm import SVR
from xgboost import XGBRegressor
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# 데이터셋 로드
file_name = '0. K-net_datasets for RF_v6.xlsx'
data = pd.read_excel(file_name)
data.columns = data.columns.str.strip()  # 열 이름에서 공백 제거

# 특정 조건에 맞지 않는 데이터 제거 ('SG1' 또는 'Out_Kwak'이 1인 데이터 제외)
data = data[(data['SG'] != 'SG1') & (data['Out_Kwak'] != 1)].reset_index(drop=True)

# 사용할 특성(피처) 및 목표 변수 지정
numerical_features = ['avg_n_value', 'depth']
binary_encoded_features = ['soil_column_1', 'soil_column_2', 'soil_column_3', 'soil_column_4']
features = numerical_features + binary_encoded_features  # 특성 목록
target = 'Vs_measured'  # 목표 변수

# 특성(X)와 목표 변수(y) 준비
X = data[features]
y = data[target]

# IQR(사분위 범위)을 이용한 이상치 탐지 및 제거
data_combined = pd.concat([X, y], axis=1)  # 특성과 목표 변수를 결합하여 하나의 데이터프레임 생성
for column in data_combined.columns:
    Q1 = data_combined[column].quantile(0.3)  # 1사분위수
    Q3 = data_combined[column].quantile(0.7)  # 3사분위수
    IQR = Q3 - Q1  # IQR 계산
    lower_bound = Q1 - 1.5 * IQR * 1.6  # 하한 경계
    upper_bound = Q3 + 1.5 * IQR * 1.16  # 상한 경계
    data_combined = data_combined[(data_combined[column] >= lower_bound) & (data_combined[column] <= upper_bound)]  # 이상치 제거

# 이상치를 제거한 후 다시 특성과 목표 변수로 분리
X = data_combined[features]
y = data_combined[target]

# 특성(피처) 표준화 (SVR 모델에 중요)
scaler = StandardScaler()  # 표준화 객체 생성
X_scaled = scaler.fit_transform(X)  # 특성 데이터 표준화

# 훈련 데이터와 테스트 데이터로 분리 (80% 훈련, 20% 테스트)
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# 스태킹 앙상블 모델의 기본 모델들 정의
base_models = [
    ('xgb', XGBRegressor(  # XGBoost 회귀 모델
        colsample_bytree=1.0,
        gamma=0.1,
        learning_rate=0.13,
        max_depth=7,
        n_estimators=150,
        subsample=1.0,
        random_state=42
    )),
    ('svr', SVR(kernel='rbf', C=1.0, epsilon=0.5))  # SVR(서포트 벡터 회귀) 모델
]
print(f"base_models= {base_models}")

# 스태킹 앙상블 모델 정의 (기본 모델들을 조합하여 최종 예측)
stacking_model = StackingRegressor(
    estimators=base_models,  # 기본 모델들
    final_estimator=Ridge()  # 최종 예측 모델로 Ridge 회귀 사용
)

# 스태킹 모델 훈련
stacking_model.fit(X_train, y_train)

# 예측 수행
y_pred = stacking_model.predict(X_test)

# 모델 평가 (RMSE와 R^2 값 계산)
rmse = mean_squared_error(y_test, y_pred, squared=False)  # RMSE (Root Mean Squared Error)
r2 = r2_score(y_test, y_pred)  # R^2 (결정계수)

# 평가 결과 출력
print(f"Stacking Ensemble RMSE: {rmse:.2f}")
print(f"Stacking Ensemble R^2 Score: {r2:.2f}")
